src/densenet.jl

```diff
@@ -1,92 +1,188 @@
+"""
+function dense_bottleneck(inplanes, growth_rate)
+  inner_channels = 4 * growth_rate
+  m = Chain(conv_bn((1, 1), inplanes, inner_channels; usebias=false, rev=true)...,
+            conv_bn((3, 3), inner_channels, growth_rate; pad=1, usebias=false, rev=true)...)
+
+  SkipConnection(m, (mx, x) -> cat(x, mx; dims=3))
 end
 
-@functor Bottleneck
+"""
+    transition(inplanes, outplanes)
 
-Bottleneck(in_planes, growth_rate) = Bottleneck(Chain(BatchNorm(in_planes, relu),
-                                          Conv((1, 1), in_planes => 4growth_rate),
-                                          BatchNorm(4growth_rate, relu),
-                                          Conv((3, 3), 4growth_rate => growth_rate, pad = (1, 1))))
+Create a DenseNet transition sequence
+([reference](https://arxiv.org/abs/1608.06993)).
 
-(b::Bottleneck)(x) = cat(b.layer(x), x, dims = 3)
+# Arguments
+- `inplanes`: number of input feature maps
+- `outplanes`: number of output feature maps
+"""
+transition(inplanes, outplanes) = (conv_bn((1, 1), inplanes, outplanes; usebias=false, rev=true)...,
+                                   MeanPool((2, 2)))
 
-Transition(chs::Pair{<:Int,<:Int}) = Chain(BatchNorm(chs[1], relu),
-                                            Conv((1, 1), chs),
-                                            MeanPool((2, 2)))
+"""
+    dense_block(inplanes, growth_rate, nblock)
 
-function _make_dense_layers(block, in_planes, growth_rate, nblock)
-    local layers = []
-    for i in 1:nblock
-        push!(layers, block(in_planes, growth_rate))
-        in_planes += growth_rate
-    end
-    Chain(layers...)
-end
+Create a sequence of `nblock` DesNet bottlenecks with `growth_rate`
+([reference](https://arxiv.org/abs/1608.06993)).
 
-function _densenet(nblocks = [6, 12, 24, 16]; block = Bottleneck, growth_rate = 32, reduction = 0.5, num_classes = 1000)
-    num_planes = 2growth_rate
-    layers = []
-    push!(layers, Conv((7, 7), 3 => num_planes, stride = (2, 2), pad = (3, 3)))
-    push!(layers, BatchNorm(num_planes, relu))
-    push!(layers, MaxPool((3, 3), stride = (2, 2), pad = (1, 1)))
-
-    for i in 1:3
-        push!(layers, _make_dense_layers(block, num_planes, growth_rate, nblocks[i]))
-        num_planes += nblocks[i] * growth_rate
-        out_planes = Int(floor(num_planes * reduction))
-        push!(layers, Transition(num_planes => out_planes))
-        num_planes = out_planes
-    end
-
-    push!(layers, _make_dense_layers(block, num_planes, growth_rate, nblocks[4]))
-    num_planes += nblocks[4] * growth_rate
-    push!(layers, BatchNorm(num_planes, relu))
-
-    Chain(layers..., MeanPool((7, 7)),
-        x->reshape(x, :, size(x, 4)),
-        Dense(num_planes, num_classes), softmax)
+# Arguments
+- `inplanes`: number of input feature maps to the full sequence
+- `growth_rate`: the rate at which output feature maps grow across blocks
+- `nblock`: the number of blocks
+"""
+function dense_block(inplanes, growth_rate, nblock)
```

Comment:
To me its not obvious how `growth_rate` should be expressed - I'd almost expect it to work like gain and multiply. Maybe just accept a vector of values and drop `nblock`?

Reply:
This is of course something that would be needed in more places than just here.

Reply:
One of the goals was to match the paper as closely as possible. So `growth_rate` corresponds to terminology in the paper (also defined as "k" in the paper which is mentioned in the docstrings).

We could combine `nblock` and `growth_rate` into `nout_channels` as a vector which I agree makes most sense. Maybe there can be a convenience method to make this vector given `nblock` and `growth_rate`.